util/stop-handler/src/stop_register.rs

```diff
@@ -0,0 +1,78 @@
+use ckb_channel::TrySendError;
+use ckb_logger::{debug, error, info, trace, warn};
+use ckb_util::Mutex;
+use tokio_util::sync::CancellationToken;
+
+struct CkbServiceHandles {
+    thread_handles: Vec<(String, std::thread::JoinHandle<()>)>,
+}
+
+/// Wait all ckb services exit
+pub fn wait_all_ckb_services_exit() {
+    info!("waiting exit signal...");
+    let exit_signal = new_crossbeam_exit_rx();
+    let _ = exit_signal.recv();
+    debug!("received exit signal, broadcasting exit signal to all threads");
+    let mut handles = CKB_HANDLES.lock();
+    for (name, join_handle) in handles.thread_handles.drain(..) {
+        match join_handle.join() {
+            Ok(_) => {
+                debug!("wait thread {} done", name);
+            }
+            Err(e) => {
+                warn!("wait thread {}: ERROR: {:?}", name, e)
```

Comment:
Did you verify the error log contains panic message?
https://stackoverflow.com/questions/69387239/what-is-in-err-when-i-try-to-join-a-panicked-thread

Reply:
This WARN log didn't print thread's panic stacktrace, I should use:
```rust
    warn!("wait thread {}: ERROR: {:?}",name,e.downcast_ref::<String>())
```

Reply:
Update:
If the `panic!` use a formatter, `e.downcast_ref::<&str>()` will get `None`, so I have to use `e.downcast_ref::<String>()`:
```rust
                let msg = e.downcast_ref::<str>().unwrap_or_else(|| {
                    e.downcast_ref::<String>()
                        .map(|s| s.as_str())
                        .unwrap_or("downcast_ref didn't get any error")
                });
                warn!("wait thread {}: ERROR: {:?}", name, msg)
```